### PR TITLE
[BEAM-2371] Port DirectRunner override matchers to language-agnostic construction APIs

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

R: @tgroh

### CODE REVIEW

**Comment by kennknowles** [comment]
retest this please

**Comment by kennknowles** [comment]
I've done local `mvn clean install` to refresh myself, and the SDF matchers are not properly in place.

**Comment by kennknowles** [comment]
Fixed SDF matching.

**Comment by kennknowles** [comment]
In topological order, this PR adds additional translation support for each requisite transform, then adds adequate `PTransformMatchers` that leverage the new support, then alters the `DirectRunner` to use them.

**Comment by kennknowles** [comment]
Locations of various methods and fields TBD and I'm happy to oblige suggestions.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java [hunk 1]
```diff
@@ -150,38 +189,44 @@ FunctionSpec translate(AppliedPTransform<?, ?, T> application, SdkComponents com
...
-          InputT extends PInput, OutputT extends POutput, PayloadT extends Message>
+          InputT extends PInput, OutputT extends POutput>
       extends PTransform<InputT, OutputT> {
 
+    @Nullable
     public abstract String getUrn();
 
     @Nullable
-    PayloadT getPayload() {
+    public Any getPayload() {
```
Comment (tgroh) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java]: This use of Any makes me very suspicious that we aren't using `Any` super well

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java [hunk 2]
```diff
@@ -496,6 +497,25 @@ private static SdkFunctionSpec toProto(ViewFn<?, ?> viewFn) {
         .build();
   }
 
+  private static <T> ParDoPayload getParDoPayload(AppliedPTransform<?, ?, ?> transform)
+      throws IOException {
+    return PTransformTranslation.toProto(
```
Comment (tgroh) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java]: This feels really circuitous. I think I understand why you're doing it, but a comment probably wouldn't be out of place here.

Comment (kennknowles) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java]: It is definitely circuitous. I wanted to support direct `Pipeline`-to-runner interactions as well as rehydrated pipelines. There are a couple other options:

1. Don't support that. Make every runner operate only on rehydrated `Pipeline` filled only with `RawPTrasnform`. That is probably a win for portability, but somewhat difficult in transition.
2. Get rid of all the static methods for pulling stuff out, and just have a `toRawPTransform` method; let runners shuffle proto fields. Eventually they may need to anyhow.
3. Add wrapper `RawPTransform` subclasses for each of the `XYZTranslation` classes with all the same methods that are currently static methods. These could be the result of dehydration or wrap a native Java class, so you still get the direct-to-runner capability.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java (hunk 1 above)
Comment (kennknowles) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java]: Acknowledged. I support your notion of making this `bytes`.

**Comment by kennknowles** [comment]
Peeled off #3274 for `WriteFiles` translation.

**Comment by kennknowles** [comment]
To reflect the bits that have been broken out into #3281, #3282, and #3283, I have removed those commits from this PR. The tests will no longer pass until the prerequisites are resolved.

**Comment by kennknowles** [comment]
Rebased on the fleshed out translations.

**Comment by kennknowles** [comment]
Build looks like a maven success + a Jenkins failure

**Comment by kennknowles** [comment]
retest this please

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/PTransformMatchersTest.java [hunk 3]
```diff
@@ -272,6 +279,18 @@ public void parDoMultiSplittable() {
   }
 
   @Test
+  public void parDoSplittable() {
+    AppliedPTransform<?, ?, ?> parDoApplication =
+        getAppliedTransform(
+            ParDo.of(splittableDoFn).withOutputTags(new TupleTag<Integer>(), TupleTagList.empty()));
+    assertThat(PTransformMatchers.splittableParDo().matches(parDoApplication), is(true));
+
+    assertThat(PTransformMatchers.stateOrTimerParDoMulti().matches(parDoApplication), is(false));
```
Comment (tgroh) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/PTransformMatchersTest.java]: Should there be a test for `parDoWithState` as well?

Comment (kennknowles) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/PTransformMatchersTest.java]: Done

**Pull request merged**